### ISSUE TITLE
Replay dialog with special commands support (like !quit)

### ISSUE THREAD

**Issue by RonWang** [alexa/ask-cli#56]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[ ] Performance issue
[x] Feature request
[ ] Documentation issue or request
[ ] Other... Please describe:
</code></pre>

This is a feature request originally reported from https://github.com/alexa/alexa-skills-kit-sdk-for-nodejs/issues/505. 

A brief description of the issue is: because the `stopIntent` doesn't promise the closure of skill session, supporting the  special commands like !quit for --replay  empowers users to control the lifecycle of skill session. This will further supports skill developers to manage session for each test case for integration test.

## Expected Behavior

If ask-cli-x supports "dialog" command, and the same --reply feature, special commands should be supported.

**Comment by pbheemag** [comment]
Dialog command needs to switch to Interactive mode once replaying all userInputs is done in replay mode. If we append `.quit` to userInputs array in replay file, this feature cannot be attained since `.quit` command forces CLI to exit the Dialog command process.
Furthermore, `.quit` command can be either manually added to userInputs array in replay file or through a script to support integration tests. 
Closing this issue.

**Comment by reflexdemon** [comment]
> Dialog command needs to switch to Interactive mode once replaying all userInputs is done in replay mode

Is it a hard requirement? This does not make sense. When we try to automate the pipelines to run integration tests it fails the tests and that is creating unwanted noise brweeen the teams.

I am not satisfied with the response. 👎

**Comment by pbheemag** [comment]
Hi @reflexdemon , 
We regret that you were not satisfied with the response.

**Is it a hard requirement?** 
One major use case I can think of is, user interacting with a skill, goes through multiple rounds of conversations (20+) and records them in a history file. Once the user replays the file, we'd like to prompt user to continue with conversation from where they left off rather than from the beginning. Hence the switching to interaction mode after replay mode.

  If we append .quit to the list of utterances in the replay file, this cannot be achieved. Also, we'd not want to add anything to replay file apart from what the user is expecting to be added. 

Could you kindly explain your use case (probably how difficult it is to edit replay file for your testing purposes) so that we can better assist you with that? Thank you.

**Comment by btburton42** [comment]
> If we append .quit to the list of utterances in the replay file, this cannot be achieved. Also, we'd not want to add anything to replay file apart from what the user is expecting to be added.

@pbheemag I believe there is some confusion about the ask from developers like @reflexdemon and myself, though I can only speak for my needs.

I am looking for **the ability** to add `!quit` to my replay file so that I can use this functionality in end-to-end automation tests. I am not looking for the script to append `!quit` to every replay session, just for the script to support the `!quit` functionality. 

Adding **support** for special commands would preserve the interactivity for those looking to use it, as you mention, and also allow for developers like myself to create end to end automation without having to rely on hacks like observing non-zero exit codes in node child processes to achieve this. 

I hope this clarifies the issue.

**Comment by pbheemag** [comment]
Hi @btburton42 @reflexdemon ,
Firstly, thank you for the clarification. We've added a flag to the record command, `Usage: .record <fileName> --append-quit` which will add `.quit` to list of utterances which are to be cached in replay file. Kindly let me know if that resolves the issue. Please refer to documentation [here](https://github.com/alexa-labs/ask-cli/blob/develop/docs/concepts/Dialog-Command.md#special-sub-commands).

**Comment by btburton42** [comment]
Hi there @pbheemag, I've taken a look at the updates to ask-cli 2.0 and I have some concerns.

1. The `--append-quit` flag does not seem to be present. Is this intended for a future release? Also, what do you mean by use the `.record` method? There is only `ask dialog [options]`
2. On that note, it also look like the `-o` flag has been removed and now there is no way to pipe dialog output to a file. We were relying on this for dialog metadata? Should I open a new issue for that?
3. Adding `!quit` to a reply file does seem to work now. Which is some good news.

**Comment by reflexdemon** [comment]
> > If we append .quit to the list of utterances in the replay file, this cannot be achieved. Also, we'd not want to add anything to replay file apart from what the user is expecting to be added.
> 
> @pbheemag I believe there is some confusion about the ask from developers like @reflexdemon and myself, though I can only speak for my needs.
> 
> I am looking for **the ability** to add `!quit` to my replay file so that I can use this functionality in end-to-end automation tests. I am not looking for the script to append `!quit` to every replay session, just for the script to support the `!quit` functionality.
> 
> Adding **support** for special commands would preserve the interactivity for those looking to use it, as you mention, and also allow for developers like myself to create end to end automation without having to rely on hacks like observing non-zero exit codes in node child processes to achieve this.
> 
> I hope this clarifies the issue.

@btburton42  thanks for detailing on my behalf. I agree with you 100 on your comment.

**Comment by reflexdemon** [comment]
> Hi there @pbheemag, I've taken a look at the updates to ask-cli 2.0 and I have some concerns.
> 
>     1. The `--append-quit` flag does not seem to be present. Is this intended for a future release? Also, what do you mean by use the `.record` method? There is only `ask dialog [options]`
> 
>     2. On that note, it also looks like the `-o` flag has been removed from `ask dialog` and now there is no way to pipe dialog output to a file. We were relying on this for parsing dialog metadata. Should I open a new issue for that?
> 
>     3. Adding `!quit` to a replay file does seem to work now. Which is some good news.

This is good and concerning. I like the good news but not happy with the other new issues. the `-o` option is something I count on also.

**Comment by btburton42** [comment]
> 1. The `--append-quit` flag does not seem to be present. Is this intended for a future release? Also, what do you mean by use the `.record` method? There is only `ask dialog [options]`
> 2. On that note, it also looks like the `-o` flag has been removed from `ask dialog` and now there is no way to pipe dialog output to a file. We were relying on this for parsing dialog metadata. Should I open a new issue for that?
> 3. Adding `!quit` to a replay file does seem to work now. Which is some good news. 


I am seeing No. 1 detailed in the launch boilerplate when you launch `ask dialog`. Now it is really just No. 2 above that is concerning.

**Comment by RonWang** [comment]
Hi @reflexdemon and @btburton42 ,

Thanks for your feedbacks! I'm really interested and eager to know more about how you are using the `--output` result for your usage, and we would like to learn from your usage to improve the dialog command experiences. We removed the `--output` option because we were thinking its information is the subset of what `--debug` shows to you. 

As of now, you might already know the current `--output` is displaying the JSON response (plain text, not a structured file format) from `get-simulation` API request, only when that simulation is successful or failed. Thus, as an alternate solution, can `--debug`, which displays the full network request response, help for now?

Thanks,
Ron

**Comment by RonWang** [comment]
Close this issue because of the inactivity. Please feel free re-open it if the solution above doesn't solve your problem, and we will also consider to enhance the experiences over the "dialog" command, including the exposure of the key request-response.

**Comment by btburton42** [comment]
@RonWang I have been away from my project this week so I haven't had a chance to test the `--debug` flag. I will do so next week and report back.

As far as the output command was concerned, it would be great to receive back proper JSON so that we'd wouldnt have to parse stdout looking for the meaningful parts.
